examples/step-6/doc/intro.dox

```diff
@@ -167,6 +175,49 @@ module that explains how these constraints can be computed and what classes in
 deal.II work on them.
 
 
+<h3> How to deal with hanging nodes in practice </h3>
+
+The practice of hanging node constraints is rather simpler than the
+theory we have outlined above. In reality, you will really only have to
+add about half a dozen lines of additional code to a program like step-4
+to make it work with adaptive meshes that have hanging nodes. The
+interesting part about this is that it is entirely independent of the
+equation you are solving: How the constraints look like has nothing
+to do with the equation and only with the finite element in use.
+As a consequence, the code to deal with these constraints is entirely
+contained in the deal.II library itself, and you do not need to worry
+about the details.
+
+The steps you have to do to make this work are essentially like this:
+- You have to have a ConstraintMatrix object that stores the
+  constraints due our desire to keep the solution space continuous
+  even in the presence of hanging nodes.
+- You have to fill this object using the function
+  DoFTools::make_hanging_node_constraints().
+- You have to use this object when you copy the local contributions to
+  the matrix and right hand side into the global objects, by using
+  ConstraintMatrix::distribute_local_to_global(). Up until
+  now, we have done this ourselves, but now with constraints, this
+  is where the magic happens and we apply the constraints to the
+  linear system. What this function does is make sure that the
+  degrees of freedom located at hanging nodes are not, in fact,
+  really free. Rather, they are factually eliminated from the
+  linear system by setting their rows and columns to zero and putting
+  something on the diagonal to ensure the matrix remains invertible.
```

Comment:
It might be worth adding a note explaining that since we use Gaussian elimination the resulting matrix (for the Laplace equation) will still be positive definite. It took me awhile when I got started with the library to realize that this is why we could not recycle condensed matrices without also fixing the right-hand side.

Reply:
I added a sentence, but this is a much longer discussion to explain correctly. It took me many years before I could fully explain what the algorithm really does (despite the fact that I implemented it myself), and I think it's too big a topic for this introduction. I'd rather stick with "some magic happens" and "the result is a SPD matrix" than try to do this topic justice.

Reply:
Fair enough.

Reply:
One thing that help me understand constraints is the matrix form. Since we constrain the space of admissible solutions and test functions by a transformation `u -> C u + f`, upon insertion into the matrix system we get `C^T A C` for the system matrix and some contribution `C^T b - C^T A f` for the right hand side. Note that one ignores the inhomogeneities in the test function slot, which is why the contribution from the solution slot shows up here.

This approach can be derived from an extended system of saddle point form and is used in a more general context as well by other projects; the case that we call 'we cannot work with this constraint object because there was a cycle in the constraints' can also be handled by solving some auxiliary linear system using saddle point techniques. Due to the construction from the elimination out of a saddle point system, the above system `C^T A C` is the reduced system without hanging nodes conceptually, so the explanation of what happens in our code is not fully complete with these formulas only as we also include a diagonal matrix for the constrained entries.

Reply:
I would believe that it would not be too complicated to discuss these formulas at this point, and some people are really helped in their understanding by a simple formula, but I can also see the point that you want things to be compact here. What do others think?

Reply:
I think a better place for this sort of discussion may be the "constraints" module, or the class documentation of the `ConstraintMatrix` (errr: `AffineConstraints`) class.

The point is that the derivation of the linear system we produce really is not trivial. By the way isn't the matrix `Z^T A Z` where `Z` is a basis for the null space? I keep forgetting -- which supports my point that it's not trivial. It turns out that it's not necessary to understand in detail either: I bet the 90% of the users of constraints have no idea how this actually works.

Reply:
In my formula, the matrix `C` is of the matrix which contains `1` on the *diagonal* [1] for unconstrained entries and the usual expansion to the constrained entries, i.e., the expression of `u_i` by some `u_j` entries in terms of the constraints. So yes, it is the basis for the null space in the actual matrix describing the constraints as `C u + f = 0`, and we are talking about the same thing, I just used a different notation for what `C` is.

I agree that the constraints module would be an equally good or even better place - whatever we think is the appropriate place. I see that we already have the appropriate formulas for the general case in the section on [Applying constraints indirectly with a LinearOperator](https://www.dealii.org/developer/doxygen/deal.II/group__constraints.html) section, so we just need to look through that at some point and see if we can spell out that formula already earlier when we describe the process that happens with matrices.

[1] It is not really the diagonal because `C` is rectangular, so it would be more precise to say the location in `C` between the same DoF in terms of the unconstrained and constrained solution space.